pkg/rehearse/executor.go

```diff
@@ -0,0 +1,140 @@
+	dryRun     bool
+	rehearsals []*prowconfig.Presubmit
+	prNumber   int
+	prRepo     string
+	refs       *pjapi.Refs
+	loggers    Loggers
+	pjclient   pj.ProwJobInterface
+}
+
+func NewExecutor(toBeRehearsed map[string][]prowconfig.Presubmit, prNumber int, prRepo string, refs *pjapi.Refs,
+	dryRun bool, loggers Loggers, pjclient pj.ProwJobInterface) *Executor {
+	rehearsals := configureRehearsalJobs(toBeRehearsed, prRepo, prNumber, loggers)
+	return &Executor{
+		dryRun:     dryRun,
+		rehearsals: rehearsals,
+		prNumber:   prNumber,
+		prRepo:     prRepo,
+		refs:       refs,
+		loggers:    loggers,
+		pjclient:   pjclient,
+	}
+}
+
+// ExecuteJobs takes configs for a set of jobs which should be "rehearsed", and
+// creates the ProwJobs that perform the actual rehearsal. *Rehearsal* means
+// a "trial" execution of a Prow job configuration when the *job config* config
+// is changed, giving feedback to Prow config authors on how the changes of the
+// config would affect the "production" Prow jobs run on the actual target repos
+func (e *Executor) ExecuteJobs() (bool, error) {
+	submitSuccess := true
+	pjs, err := e.submitRehearsals()
+	if err != nil {
+		submitSuccess = false
+	}
+
+	if !e.dryRun {
+		if submitSuccess {
+			return true, nil
+		}
+		return true, fmt.Errorf("failed to submit all rehearsal jobs")
+	}
+	req, err := labels.NewRequirement(rehearseLabel, selection.Equals, []string{strconv.Itoa(e.prNumber)})
+	if err != nil {
+		return false, fmt.Errorf("failed to create label selector: %v", err)
+	}
+	selector := labels.NewSelector().Add(*req).String()
+
+	waitSuccess, err := e.waitForJobs(pjs, selector)
+
+	if !submitSuccess {
+		return waitSuccess, fmt.Errorf("failed to submit all rehearsal jobs")
+	}
+	return waitSuccess, err
+
+}
+
+func (e *Executor) waitForJobs(jobs sets.String, selector string) (bool, error) {
+	if len(jobs) == 0 {
+		return true, nil
+	}
+	success := true
+	for {
+		w, err := e.pjclient.Watch(metav1.ListOptions{LabelSelector: selector})
+		if err != nil {
+			return false, fmt.Errorf("failed to create watch for ProwJobs: %v", err)
+		}
+		defer w.Stop()
+		for event := range w.ResultChan() {
+			pj, ok := event.Object.(*pjapi.ProwJob)
+			if !ok {
+				return false, fmt.Errorf("received a %T from watch", event.Object)
+			}
+			fields := pjutil.ProwJobFields(pj)
+			fields["state"] = pj.Status.State
+			e.loggers.Debug.WithFields(fields).Debug("Processing ProwJob")
+			if !jobs.Has(pj.Name) {
+				continue
+			}
+			switch pj.Status.State {
+			case pjapi.FailureState, pjapi.AbortedState, pjapi.ErrorState:
+				e.loggers.Job.WithFields(fields).Error("Job failed")
+				success = false
+			case pjapi.SuccessState:
+				e.loggers.Job.WithFields(fields).Info("Job succeeded")
+			default:
+				continue
+			}
+			jobs.Delete(pj.Name)
+			if jobs.Len() == 0 {
+				return success, nil
+			}
+		}
+	}
+}
+
+func (e *Executor) submitRehearsals() (sets.String, error) {
+	var validationErrors []error
```

Comment:
Why is this called `validationErrors`? This is doing no validation.

Reply:
addressed